marimo/_plugins/ui/_impl/dataframes/handlers.py

```diff
@@ -0,0 +1,186 @@
+            elif condition.operator == "is_true":
+                df_filter = df[condition.column_id] is True
+            elif condition.operator == "is_false":
+                df_filter = df[condition.column_id] is False
+            elif condition.operator == "is_nan":
+                df_filter = df[condition.column_id].isna()
+            elif condition.operator == "is_not_nan":
+                df_filter = df[condition.column_id].notna()
+            elif condition.operator == "equals":
+                df_filter = df[condition.column_id].eq(value)
+            elif condition.operator == "contains":
+                df_filter = df[condition.column_id].str.contains(
+                    value, regex=False
+                )
+            elif condition.operator == "regex":
+                df_filter = df[condition.column_id].str.contains(
+                    value, regex=True
+                )
+            elif condition.operator == "starts_with":
+                df_filter = df[condition.column_id].str.startswith(value)
+            elif condition.operator == "ends_with":
+                df_filter = df[condition.column_id].str.endswith(value)
+            elif condition.operator == "in":
+                df_filter = df[condition.column_id].isin(value)
+            else:
+                _assert_never(condition.operator)
+
+            if transform.operation == "keep_rows":
+                df = df[df_filter]
+            elif transform.operation == "remove_rows":
+                df = df[~df_filter]
+            else:
+                _assert_never(transform.operation)
+        return df
+
+    @staticmethod
+    def handle_group_by(
+        df: "pd.DataFrame", transform: GroupByTransform
+    ) -> "pd.DataFrame":
+        group = df.groupby(transform.column_ids, dropna=transform.drop_na)
+        if transform.aggregation == "count":
+            return group.count()
+        elif transform.aggregation == "sum":
+            return group.sum()
+        elif transform.aggregation == "mean":
+            return group.mean()
+        elif transform.aggregation == "median":
+            return group.median()
+        elif transform.aggregation == "min":
+            return group.min()
+        elif transform.aggregation == "max":
+            return group.max()
+        else:
+            _assert_never(transform.aggregation)
+
+    @staticmethod
+    def handle_aggregate(
+        df: "pd.DataFrame", transform: AggregateTransform
+    ) -> "pd.DataFrame":
+        dict_of_aggs = {
+            column_id: transform.aggregations
+            for column_id in transform.column_ids
+        }
+        return df.agg(dict_of_aggs)  # type: ignore[arg-type]
+
+
+def apply_transforms(
+    df_prev: "pd.DataFrame", transforms: Transformations
+) -> "pd.DataFrame":
+    if transforms.transforms is None or len(transforms.transforms) == 0:
+        return df_prev
+    # defensive copy
+    df = df_prev.copy()
```

Comment:
Can we remove this copy? Pandas operations aren't in-place by default, ie they already make a copy.